### PR TITLE
Add `product_search_space` function.

### PR DESCRIPTION
**Pull request by sile**
This PR adds  `optuna.samplers.product_search_space` function.

See also: https://github.com/pfnet/optuna/pull/399

### CODE REVIEW

File: optuna/study.py [hunk 1]
```diff
@@ -578,6 +578,41 @@ def trials(self):
 
         return self.storage.get_all_trials(self.study_id)
 
+    @property
+    def product_search_space(self):
+        # type: () -> Dict[str, BaseDistribution]
+        """Return the product search space of the :class:`~optuna.study.InTrialStudy`.
+
+        "product search space" contains the product of parameter distributions that have been
```
Comment (toshihikoyanase) [inline optuna/study.py]: IMO, the expression 'product of parameter distributions' seems to be confusing with [product distribution](https://en.wikipedia.org/wiki/Product_distribution). Can we rephrase this as something like 'product set of parameter distributions'?

File: optuna/study.py [hunk 2]
```diff
@@ -578,6 +578,41 @@ def trials(self):
...
+        suggested in the completed trials of the study so far.
+        If there are two parameters that have the same name but different distributions, neither
+        is included in the resulting search space (i.e., dynamic value space isn't supported).
```
Comment (toshihikoyanase) [inline optuna/study.py]: > "dynamic value space isn't supported" 

I think some users may think that dynamic value space cause exceptions. Can we describe the behavior directly like 'The parameters with dynamic value ranges will be excluded.' or something?

File: optuna/study.py (hunk 1 above)
Comment (sile) [reply optuna/study.py]: That makes sense. Thank you.

File: optuna/study.py (hunk 2 above)
Comment (sile) [reply optuna/study.py]: > 'The parameters with dynamic value ranges will be excluded.'

This phrase seems good.

File: tests/samplers_tests/test_samplers.py [hunk 3]
```diff
@@ -201,3 +201,43 @@ def objective(trial):
...
+        trial.suggest_uniform('z', 0, 1)
+        raise exception
+
+    study.optimize(lambda t: objective(t, RuntimeError()), n_trials=1)
+    study.optimize(lambda t: objective(t, optuna.structs.TrialPruned()), n_trials=1)
+    assert optuna.samplers.product_search_space(study) == {
+        'y': UniformDistribution(low=-3, high=3)
+    }
+
+    # If two trials have the same name but different distributions,
```
Comment (g-votte) [inline tests/samplers_tests/test_samplers.py]: Please replace `trials` with `parameters`.

Comment (sile) [reply tests/samplers_tests/test_samplers.py]: Oops.., thanks!

Fixed: https://github.com/pfnet/optuna/pull/430/commits/49b01d6a0cffae62b5c0e78190e248b3fcd66f6d
